src/util/obbt.jl

```diff
@@ -0,0 +1,406 @@
+                gen["cost"][3]
+            for (i,gen) in ref(pm, :gen)) +
+            sum(   
+                dcline["cost"][1]*var(pm,:p_dc, from_idx[i])^2 +
+                dcline["cost"][2]*var(pm, :p_dc, from_idx[i])  +
+                dcline["cost"][3] 
+            for (i,dcline) in ref(pm, :dcline))
+            <= bound 
+    )
+end 
+
+function constraint_obj_bound(pm::GenericPowerModel, bound)
+    model = PowerModels.check_cost_models(pm)
+    if model != 2 
+        error("Only cost models of type 2 is supported at this time, given cost model type $(model)")
+    end 
+    constraint_obj_bound_polynomial(pm, bound)
+end 
+
+function create_modifications(pm::GenericPowerModel, 
+    vm_lb::Dict{Any,Float64}, vm_ub::Dict{Any,Float64}, 
+    td_lb::Dict{Any,Float64}, td_ub::Dict{Any,Float64}) 
+    
+    modifications = Dict{String,Any}()
+
+    modifications["per_unit"] = true
+    modifications["bus"] = Dict{String,Any}()
+    modifications["branch"] = Dict{String,Any}()
+
+    buses = ids(pm, :bus)
+    buspairs = ids(pm, :buspairs)
+    branches = ids(pm, :branch)
+
+    for bus in buses
+        index = string(ref(pm, :bus, bus, "index"))
+        modifications["bus"][index] = Dict{String,Any}( "vmin" => vm_lb[bus], "vmax" => vm_ub[bus] )
+    end 
+
+    for branch in branches 
+        index = string(ref(pm, :branch, branch, "index"))
+        f_bus = ref(pm, :branch, branch, "f_bus") 
+        t_bus = ref(pm, :branch, branch, "t_bus")
+        bp = (f_bus, t_bus)
+        modifications["branch"][index] = Dict{String,Any}( "angmin" => td_lb[bp], "angmax" => td_ub[bp] )
+    end 
+
+    return modifications
+end 
+
+function create_modifications(pm::GenericPowerModel, lb::Float64, ub::Float64, id::Int64)
+    
+    modifications = Dict{String,Any}()
+
+    modifications["per_unit"] = true
+    modifications["bus"] = Dict{String,Any}()
+
+    index = string(ref(pm, :bus, id, "index"))
+    modifications["bus"][index] = Dict{String,Any}( "vmin" => lb, "vmax" => ub )
+
+    return modifications
+
+end 
+
+function create_modifications(pm::GenericPowerModel, lb::Float64, ub::Float64, id::Tuple{Int64,Int64})
+
+    modifications = Dict{String,Any}()
+
+    modifications["per_unit"] = true
+    modifications["branch"] = Dict{String,Any}()
+
+    branches = ids(pm, :branch)
+
+
+    for branch in branches 
+        index = string(ref(pm, :branch, branch, "index"))
+        f_bus = ref(pm, :branch, branch, "f_bus") 
+        t_bus = ref(pm, :branch, branch, "t_bus")
+        bp = (f_bus, t_bus)
+        if bp == id 
+            modifications["branch"][index] = Dict{String,Any}( "angmin" => lb, "angmax" => ub )
+        end
+    end 
+
+    return modifications
+
+end
+
+function run_obbt_opf(file::String, model_constructor, solver; kwargs...)
+    data = PowerModels.parse_file(file)
+    return run_obbt_opf(data, model_constructor, solver; kwargs...)
+end 
+
+function run_obbt_opf(data::Dict{String,Any}, model_constructor, solver;
```

Comment:
I would recommend having a recommended default model constructor.  I think the average user just wants the bounds, not too much about how they are calculated.

Reply:
done, model_constructor is now a keyword argument